streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -16,78 +16,408 @@
      * @param applicationState        the metadata and other info describing the current application state
      *
-     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default
-     *         rack-aware assignment for standby tasks
+     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default rack-aware assignment for standy tasks
      */
     public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTasks(
         final ApplicationState applicationState,
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments
     ) {
+        if (!hasValidRackInformation(applicationState)) {
+            LOG.warn("Cannot optimize standby tasks with invalid rack information.");
+            return kafkaStreamsAssignments;
+        }
+
+        final int crossRackTrafficCost = applicationState.assignmentConfigs().rackAwareTrafficCost();
+        final int nonOverlapCost = applicationState.assignmentConfigs().rackAwareNonOverlapCost();
+        final long currentCost = computeTaskCost(
+            applicationState.allTasks(),
+            applicationState.kafkaStreamsStates(false),
+            crossRackTrafficCost,
+            nonOverlapCost,
+            true,
+            true
+        );
+        LOG.info("Assignment before standby task optimization has cost {}", currentCost);
         throw new UnsupportedOperationException("Not Implemented.");
     }
 
+    private static long computeTaskCost(final Set<TaskInfo> tasks,
+                                        final Map<ProcessId, KafkaStreamsState> clients,
+                                        final int crossRackTrafficCost,
+                                        final int nonOverlapCost,
+                                        final boolean hasReplica,
+                                        final boolean isStandby) {
+        if (tasks.isEmpty()) {
+            return 0;
+        }
+
+        final List<UUID> clientIds = clients.keySet().stream().map(ProcessId::id).collect(
+            Collectors.toList());
+
+        final List<TaskId> taskIds = tasks.stream().map(TaskInfo::id).collect(Collectors.toList());
+        final Map<TaskId, Set<TaskTopicPartition>> topicPartitionsByTaskId = tasks.stream().collect(
+            Collectors.toMap(TaskInfo::id, TaskInfo::topicPartitions));
+
+        final Map<UUID, Optional<String>> clientRacks = clients.values().stream().collect(
+            Collectors.toMap(state -> state.processId().id(), KafkaStreamsState::rackId));
+
+        final Map<UUID, Set<TaskId>> taskIdsByProcess = clients.values().stream().collect(
+            Collectors.toMap(state -> state.processId().id(), state -> {
+                if (isStandby) {
+                    return state.previousStandbyTasks();
+                }
+                return state.previousActiveTasks();
```

Comment:
this is what we're passing in that ultimately becomes the `hasAssignedTask` predicate for the graph constructors, right? We actually need the current assignment (so what's in the KafkaStreamsAssignment) not the previous assignment (ie what's in the KafkaStreamsState). See https://github.com/apache/kafka/pull/16033/files#r1618135137

Reply:
Ah I just reread my [original comment](https://github.com/apache/kafka/pull/16033#discussion_r1618282249) about this and I think it's my fault for poorly wording what I was trying to actually suggest there. Anyways this should all be cleared up as part of the POC so hopefully we don't have to think about this ever again 😄